ros/src/computing/planning/motion/packages/waypoint_planner/nodes/astar_avoid/astar_avoid.cpp

```diff
@@ -0,0 +1,407 @@
+{
+  current_pose_global_ = msg;
+
+  if (!enable_avoidance_)
+  {
+    current_pose_initialized_ = true;
+  }
+  else
+  {
+    current_pose_local_.pose = transformPose(
+        current_pose_global_.pose, getTransform(costmap_.header.frame_id, current_pose_global_.header.frame_id));
+    current_pose_local_.header.frame_id = costmap_.header.frame_id;
+    current_pose_local_.header.stamp = current_pose_global_.header.stamp;
+    current_pose_initialized_ = true;
+  }
+}
+
+void AstarAvoid::currentVelocityCallback(const geometry_msgs::TwistStamped& msg)
+{
+  current_velocity_ = msg;
+  current_velocity_initialized_ = true;
+}
+
+void AstarAvoid::baseWaypointsCallback(const autoware_msgs::Lane& msg)
+{
+  base_waypoints_ = msg;
+  base_waypoints_initialized_ = true;
+}
+
+void AstarAvoid::closestWaypointCallback(const std_msgs::Int32& msg)
+{
+  closest_waypoint_index_ = msg.data;
+  closest_waypoint_initialized_ = true;
+}
+
+void AstarAvoid::obstacleWaypointCallback(const std_msgs::Int32& msg)
+{
+  obstacle_waypoint_index_ = msg.data;
+}
+
+void AstarAvoid::run()
+{
+  // check topics
+  state_ = AstarAvoid::STATE::INITIALIZING;
+
+  while (ros::ok())
+  {
+    ros::spinOnce();
+    if (checkInitialized())
+    {
+      obstacle_waypoint_index_ = -1;  // reset once
+      state_ = AstarAvoid::STATE::RELAYING;
+      break;
+    }
+    ROS_WARN("Waiting for subscribing topics...");
+    ros::Duration(1.0).sleep();
+  }
+
+  // main loop
+  int end_of_avoid_index = -1;
+  ros::WallTime start_plan_time = ros::WallTime::now();
+  ros::WallTime start_avoid_time = ros::WallTime::now();
+  autoware_msgs::Lane current_waypoints, avoid_waypoints;
+  ros::Rate rate(update_rate_);
+
+  while (ros::ok())
+  {
+    ros::spinOnce();
+
+    // relay mode
+    if (!enable_avoidance_)
+    {
+      publishWaypoints(base_waypoints_);
+      rate.sleep();
+      continue;
+    }
+
+    // avoidance mode
+    bool found_obstacle = (obstacle_waypoint_index_ >= 0);
+    bool avoid_velocity = (current_velocity_.twist.linear.x < avoid_start_velocity_ / 3.6);
+
+    // update state
+    if (state_ == AstarAvoid::STATE::RELAYING)
+    {
+      if (found_obstacle)
+      {
+        ROS_INFO("RELAYING -> STOPPING, Decelerate for stopping");
+        state_ = AstarAvoid::STATE::STOPPING;
+      }
+    }
+    else if (state_ == AstarAvoid::STATE::STOPPING)
+    {
+      if (!found_obstacle)
+      {
+        ROS_INFO("STOPPING -> RELAYING, Obstacle disappers");
+        state_ = AstarAvoid::STATE::RELAYING;
+      }
+      else if (avoid_velocity)
+      {
+        bool replan = ((ros::WallTime::now() - start_plan_time).toSec() > replan_interval_);
+        if (replan)
+        {
+          ROS_INFO("STOPPING -> PLANNING, Start A* planning");
+          state_ = AstarAvoid::STATE::PLANNING;
+          start_plan_time = ros::WallTime::now();
+          startPlanThread(current_waypoints, avoid_waypoints, end_of_avoid_index, start_avoid_time);
```

Comment:
Is it thread-safe?
Does rewriting "state_" occur at the same time?

Reply:
`state_` can be changed after checking if its transition can be executed.